### PR TITLE
add terraform to provision packet machine

### PR DESCRIPTION
**Pull request by paulfantom**
Add a script to provision packet machine and configure it so we can connect to it with libvirt client via SSH or TCP. It also automatically creates temporary ssh key file if needed (default behavior). 

/wip

### CODE REVIEW

File: hack/prebuild/main.tf [hunk 1]
```diff
@@ -0,0 +1,32 @@
+resource "packet_project" "libvirt_actuator" {
+  name = "libvirt-actuator tests"
+}
+
+resource "packet_ssh_key" "key" {
+  name       = "unlikely_tf_ssh_key_name"
```
Comment (spangenberg) [inline hack/prebuild/main.tf]: You should put ${var.environment_id} here.

File: hack/prebuild/main.tf [hunk 2]
```diff
@@ -0,0 +1,32 @@
+resource "packet_project" "libvirt_actuator" {
+  name = "libvirt-actuator tests"
```
Comment (spangenberg) [inline hack/prebuild/main.tf]: Maybe put ${var.environment_id} here.

File: hack/prebuild/main.tf [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+resource "packet_ssh_key" "key" {
+  name       = "unlikely_tf_ssh_key_name"
+  public_key = "${file("${var.ssh_key_path}")}"
+}
+
+resource "packet_device" "libvirt" {
+  hostname         = "${var.environment_id}"
```
Comment (spangenberg) [inline hack/prebuild/main.tf]: Prepend something non alpha numeric, in case uuid generates a number fist.

File: hack/prebuild/init.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+unix_sock_rw_perms = "0770"
+listen_tls = 0
+listen_tcp = 1
+auth_tcp="none"
+tcp_port = "16509"
+EOF
+
+echo 'LIBVIRTD_ARGS="--listen"' >> /etc/sysconfig/libvirtd
+
+iptables -I INPUT -p tcp --dport 16509 -j ACCEPT -m comment --comment "Allow insecure libvirt clients"
```
Comment (bison) [inline hack/prebuild/init.sh]: Are the tests going to use TCP or SSH as a transport for libvirt? I don't think it's okay to run these with unauthenticated libvirt listening on the public internet.

File: hack/prebuild/main.tf [hunk 5]
```diff
@@ -0,0 +1,32 @@
+resource "packet_project" "libvirt_actuator" {
```
Comment (bison) [inline hack/prebuild/main.tf]: Does this create a project for every build? We probably want a single re-used project. That will let us manage permissions to for it.

File: hack/prebuild/init.sh (hunk 4 above)
Comment (enxebre) [reply hack/prebuild/init.sh]: For CI I was thinking on parameterising the ip to allowed to access and then reuse your tls stuff, so we test it as it's consumed by the installer/mao. wdyt?

Comment (bison) [reply hack/prebuild/init.sh]: Yeah, that sounds good. I'm not familiar with the e2e tests or the plan for those. Do they use the installer? If so, I think this should be on-hold until https://github.com/openshift/installer/pull/296 merges.

Comment (enxebre) [reply hack/prebuild/init.sh]: Eventually we might use installer. Currently for mao/aws-actuator we use terraform to create the minimal infra needed to satisfy some actuator assumptions and verify expected behaviour, 
See https://github.com/openshift/machine-api-operator/tree/master/tests
e.g The minimal infra for libvirt actuator will contain a network and the ign volumes

Comment (bison) [reply hack/prebuild/init.sh]: Ah, cool. Then I think it should be pretty easy to use TLS or SSH as the transport from the start. There aren't really any changes required for the actuator. The credentials, certs or SSH keys, just need to be available and the right URI needs to be used. The libvirt library will handle the rest.

Comment (bison) [reply hack/prebuild/init.sh]: Just to add a little more detail on making TLS work: [here's a link to the docs](https://github.com/openshift/installer/blob/41a24535cb6025827ff4976650424d439e8c6661/docs/dev/libvirt-howto.md#17-configure-libvirt-to-accept-tls-connections)

We just need to generate the TLS assets somehow -- could use certtool, cfssl, or even Terraform if we don't want to use the Go tool in the installer PR. Then configure libvirtd to point to the certs and use a URI like: `qemu+tls://$IP/system?pkipath=/path/to/certs`

That should be enough and pretty secure as long as no one can access the generated certs and they are one-time use.

Comment (paulfantom) [reply hack/prebuild/init.sh]: It would be probably easier to adapt it to use SSH as this already need SSH keys to connect to instance.

Comment (bison) [reply hack/prebuild/init.sh]: Totally fine. As long as we're doing authentication and encryption somehow.

File: hack/packet-provision.sh [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+case ${1} in
+  "install")
+    ssh_path="$TF_VAR_ssh_key_path"
+    if [ "$TF_VAR_ssh_key_path" == "" ]; then
+        echo -e "\e[33mCreating temporary SSH file\e[0m"
+        ssh-keygen -t rsa -b 4096 -C "temporary packet.net key" -P "" -f "/tmp/packet_id_rsa" -q
+        ssh_path="/tmp/packet_id_rsa.pub"
+    fi
+    terraform init -input=false
+    terraform plan -input=false -out=tfplan.out && terraform apply -input=false -auto-approve tfplan.out
```
Comment (frobware) [inline hack/packet-provision.sh]: As we have `-e` this could be two separate commands.

File: hack/packet-provision.sh [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    echo -e "\e[32m"
+    echo -e "*** Your packet.net host is called ${TF_VAR_environment_id}"
+    echo -e "*** You can also access it via SSH with key located in ${ssh_path}"
+    echo -e "\e[0m"
+    ;;
+  "destroy")
+    terraform destroy -input=false -auto-approve
+    rm /tmp/packet_id_rsa* || :
```
Comment (frobware) [inline hack/packet-provision.sh]: Why not use `rm -f`, instead of `||`?

File: hack/prebuild/init.sh [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#/bin/bash
+
+yum install -y -d1 libvirt libvirt-daemon-kvm
+usermod -aG libvirt root
```
Comment (frobware) [inline hack/prebuild/init.sh]: Not sure I've looked to see how this script is used (or when) but do you not need `newgrp libvirt` here having modified the groups?

File: hack/prebuild/init.sh [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+listen_tcp = 1
+auth_tcp="none"
+tcp_port = "16509"
+EOF
+
+echo 'LIBVIRTD_ARGS="--listen"' >> /etc/sysconfig/libvirtd
+
+iptables -I INPUT -p tcp --dport 16509 -j ACCEPT -m comment --comment "Allow insecure libvirt clients"
+
+systemctl start libvirtd
```
Comment (frobware) [inline hack/prebuild/init.sh]: Do we also need `systemctl enable libvirtd`? Is there any expectation of this starting on reboot?

Comment (paulfantom) [reply hack/prebuild/init.sh]: No, those machines aren't expected to be rebooted.

File: hack/prebuild/init.sh (hunk 8 above)
Comment (paulfantom) [reply hack/prebuild/init.sh]: It is used to install libvirt on packet host the easiest way.

File: hack/packet-provision.sh (hunk 6 above)
Comment (paulfantom) [reply hack/packet-provision.sh]: Right, I will fix that in a follow-up PR

**Pull request merged**